### ISSUE TITLE
Comprehension Inference inside for loop

### ISSUE THREAD

**Issue by GBeauregard** [microsoft/pyright#2706]
**Describe the bug**
On Advent of Code 2021 day 4, I was working on recreational code minifcation and came upon pyright struggling with type inference inside a for loop in a way I couldn't explain. I thought it might be related to reaching some recursive type inference hard limit, but upon messing with it further today I'm not so sure so I'm offering it up in case it's a bug.

**To Reproduce**
I've created the following minimal example of the problem:
```python
# pyright: strict
boards = [[[2, 3], [4, 5]], [[2, 3], [4, 0]]]
called = set[int]()
for n in [0, 1, 2, 3, 4, 6]:
    called.add(n)
    remain = [[set(line) - called for line in zip(*b)] for b in boards]
    boards = [b for b, u in zip(boards, remain) if all(u)]
```

```console
Found 1 source file
/home/gregory/Documents/projects/advent-2021/python/04/blah.py
  /home/gregory/Documents/projects/advent-2021/python/04/blah.py:25:5 - error: Type of "remain" is partially unknown
    Type of "remain" is "list[list[set[Unknown]]]" (reportUnknownVariableType)
  /home/gregory/Documents/projects/advent-2021/python/04/blah.py:26:24 - error: Type of "u" is partially unknown
    Type of "u" is "list[set[Unknown]]" (reportUnknownVariableType)
  /home/gregory/Documents/projects/advent-2021/python/04/blah.py:26:56 - error: Argument type is partially unknown
    Argument corresponds to parameter "__iterable" in function "all"
    Argument type is "list[set[Unknown]]" (reportUnknownArgumentType)
  /home/gregory/Documents/projects/advent-2021/python/04/blah.py:26:41 - error: Argument type is partially unknown
    Argument corresponds to parameter "__iter2" in function "__new__"
    Argument type is "list[list[set[Unknown]]]" (reportUnknownArgumentType)
```

pyright has no problem figuring out the type if i move everything outside the for loop. This was a bit odd, but I thought it was within reason I was doing something too complicated for the typing inference. However, when I change `zip(*b)` to `b`, I get behavior that appears to be more clearly a bug:

```python
# pyright: strict
boards = [[[2, 3], [4, 5]], [[2, 3], [4, 0]]]
called = set[int]()
for n in [0, 1, 2, 3, 4, 6]:
    called.add(n)
    remain = [[set(line) - called for line in b] for b in boards]
    reveal_type(remain)
    boards = [b for b, u in zip(boards, remain) if all(u)]
```
```console
Found 1 source file
/home/gregory/Documents/projects/advent-2021/python/04/blah.py
  /home/gregory/Documents/projects/advent-2021/python/04/blah.py:27:24 - error: Type of "u" is partially unknown
    Type of "u" is "list[set[Unknown]]" (reportUnknownVariableType)
  /home/gregory/Documents/projects/advent-2021/python/04/blah.py:27:56 - error: Argument type is partially unknown
    Argument corresponds to parameter "__iterable" in function "all"
    Argument type is "list[set[Unknown]]" (reportUnknownArgumentType)
  /home/gregory/Documents/projects/advent-2021/python/04/blah.py:26:17 - info: Type of "remain" is "list[list[set[int]]]"
```

i.e., pyright seems to have figured out the type of `remain`, but then claims it's unknown when used within the boards comprehension on the next line. Is this a bug or is something subtle going on with scope or inference limits I don't understand?

**Expected behavior**
I expected the original code sample to strict type check.

**VS Code extension or command-line**
pyright 1.1.195 linux commandline

**Additional context**
The following type checks fine (removing the for loop):
```python
# pyright: strict
boards = [[[2, 3], [4, 5]], [[2, 3], [4, 0]]]
called = set[int]()
remain = [[set(line) - called for line in zip(*b)] for b in boards]
boards = [b for b, u in zip(boards, remain) if all(u)]
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.196, which I just published. It will also be included in the next release of pylance.

**Comment by GBeauregard** [comment]
I managed to break this again with an extra level of unpacking, but I'm not sure how far you want to go down this rabbit hole:

This doesn't type check:
```python
# pyright: strict
boards = [[[2,3],[4,5]],[[2,3],[4,0]]]
called = set[int]()
for n in [0,1,2,3,4,6]:
    called.add(n)
    remain = [[set(line) - called for line in (*b, *zip(*b))] for b in boards]
    boards = [b for b, u in zip(boards, remain) if all(u)]
```

But if I replace the tricky unpacking with a chain it works okay:
```python
# pyright: strict
from itertools import chain
boards = [[[2,3],[4,5]],[[2,3],[4,0]]]
called = set[int]()
for n in [0,1,2,3,4,6]:
    called.add(n)
    remain = [[set(line) - called for line in chain(b, zip(*b))] for b in boards]
    boards = [b for b, u in zip(boards, remain) if all(u)]
```

Both do work outside the for loop. The cases from before work fine

**Comment by erictraut** [comment]
You're shaking out some long-standing bugs! I'll take a look.

**Comment by erictraut** [comment]
Found it! It was effectively the same bug but in a different code path (this time dealing with the unpack operator). This will be fixed in the next release.

**Comment by GBeauregard** [comment]
:) Great, thanks.

**Comment by erictraut** [comment]
This is included in pyright 1.1.197, which I just published. It will also be included in the next release of pylance.
